src/backend/api/meals.js

```diff
@@ -1,15 +1,306 @@
-    const titles = await knex("meals").select("title");
-    response.json(titles);
+    const meals = await knex("meal");
+    const postedMeal = meals.insert(
+      {
+      title: request.body.title,
+      description: request.body.description,
+      location: request.body.location,
+      when: request.body.when,
+      max_reservations: request.body.max_reservations,
+      price: request.body.price,
+      created_date: request.body.created_date
+    }
+    )
+    response.json(postedMeal);
   } catch (error) {
     throw error;
   }
 });
 
+// put method 1
+router.put("/:id", async (request, response) => {
+  try {
+    const meals = await knex("meal").select("*");
+    const inputId = Number(request.params.id);
+    const maxIdOfMeal =Math.max( meals.map(meal => meal.id));
+    if (isNaN(inputId)) {
+      response.send("Id is not number");
+    }
+    else if (inputId > maxIdOfMeal) {
+      response.send(`the largest id is : ${maxIdOfMeal}`)
+    }
+    else {
+      const updatedMeal = meals
+        .where({ id: request.params.id })
+        .update({
+          title: "salad",
+          description: "danlish way of salad",
+          location: "Lyngby",
+          when: "2022-01-29",
+          max_reservations: 50,
+          price: 68,
+          created_date: "2021-01-01"
+        });
+      response.json(updatedMeal);
+    }
+  } catch (error) {
+    throw error;
+  }
+});
+//method 2  
+router.put("/:id", async (request, response) => {
+  try {
+    const meals = await knex("meal").select("*");
+    const inputId = Number(request.params.id);
+    const maxIdOfMeal =Math.max( meals.map(meal => meal.id));
+    if (isNaN(inputId)) {
+      response.send("Id is not number");
+    }
+    else if (inputId >maxIdOfMeal) {
+      response.send(`the largest id is : ${maxIdOfMeal}`)
+    }
+    else {
+      const updatedMeal = meals
+        .where({ id: inputId })
+        .update(request.body.title);
+      response.json(updatedMeal);
+    }
+  } catch (error) {
+    throw error;
+  }
+});
+
+// delete
+router.delete("/:id", async (request, response) => {
+  try {
+    const meals = await knex("meal").select("*");
+    const inputId = Number(request.params.id);
+    const maxIdOfMeal =Math.max( meals.map(meal => meal.id));
+    if (isNaN(inputId)) {
+      response.send("Id is not number");
+    }
+    else if (inputId > maxIdOfMeal) {
+      response.send(`the largest id is : ${maxIdOfMeal}`)
+    }
+    else {
+      const deletedMeal = meals
+        .where({ id: request.params.id })
+        .delete();
+      response.json(deletedMeal);
+    }
+  } catch (error) {
+    throw error;
+  }
+});
+
+
+//GET api/meals/  Query parameters 
+
+router.get("/", async (request, response) => {
+
+  if ('maxPrice' in request.query) {
+    try {
+      const meals = await knex("meal").select("*");
+      const maxPrice = Number(request.query.maxPrice);
+
+      if (isNaN(maxPrice)) {
```

Comment:
Same as in line 24: Try to first make any validation checks you want before fetching data from the database!

Reply:
The opposite :) First validate the input and then query!

Reply:
oh, yes. you are right. first, we have to judge the input and then query. :)